fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AuthorizerStmtVisitor.java

```diff
@@ -260,23 +277,40 @@ public Void visitDeleteStatement(DeleteStmt statement, ConnectContext context) {
         }
         Map<TableName, Relation> allTouchedTables = AnalyzerUtils.collectAllTableAndViewRelations(statement);
         allTouchedTables.remove(statement.getTableName());
-        checkSelectTableAction(context, allTouchedTables);
+        try {
+            Map<TableName, Set<String>> tableColumns = AnalyzerUtils.collectAllSelectTableColumns(statement);
+            tableColumns.remove(statement.getTableName());
+            checkCanSelectFromColumns(context, tableColumns, allTouchedTables);
+        } catch (ErrorReportException e) {
+            checkSelectTableAction(context, allTouchedTables);
+        }
         return null;
     }
 
     @Override
     public Void visitUpdateStatement(UpdateStmt statement, ConnectContext context) {
         try {
-            Authorizer.checkTableAction(context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
-                    statement.getTableName(), PrivilegeType.UPDATE);
+            Set<String> assignmentColumn = statement.getAssignments().stream()
+                    .map(columnAssignment -> columnAssignment.getColumn()).collect(Collectors.toSet());
+            Authorizer.checkColumnsAction(context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
+                    statement.getTableName(), assignmentColumn, PrivilegeType.UPDATE);
         } catch (AccessDeniedException e) {
-            AccessDeniedException.reportAccessDenied(statement.getTableName().getCatalog(),
-                    context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
-                    PrivilegeType.UPDATE.name(), ObjectType.TABLE.name(), statement.getTableName().getTbl());
+            try {
+                Authorizer.checkTableAction(context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
+                        statement.getTableName(), PrivilegeType.UPDATE);
+            } catch (AccessDeniedException exception) {
+                AccessDeniedException.reportAccessDenied(statement.getTableName().getCatalog(),
+                        context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
+                        PrivilegeType.UPDATE.name(), ObjectType.TABLE.name(), statement.getTableName().getTbl());
+            }
         }
         Map<TableName, Relation> allTouchedTables = AnalyzerUtils.collectAllTableAndViewRelations(statement);
-        allTouchedTables.remove(statement.getTableName());
-        checkSelectTableAction(context, allTouchedTables);
+        try {
+            Map<TableName, Set<String>> tableColumns = AnalyzerUtils.collectAllSelectTableColumns(statement);
+            checkCanSelectFromColumns(context, tableColumns, allTouchedTables);
+        } catch (ErrorReportException e) {
+            checkSelectTableAction(context, allTouchedTables);
+        }
```

Comment:
> At the table level, if the UPDATE permission for the same table passes verification, there is no need to check the SELECT permission for that table. However, this logic may not be appropriate in column-level scenarios. Within an UpdateStmt, the columns being updated and those involved in the subquery may not be the same, as in the statement update t set t.a=1 where t.b=2, where the user needs both UPDATE privilege for column 'a' and SELECT privilege for column 'b'. It would be better not to remove the target table in this case.

Good point, but is it better to adjust the code like this?

```suggestion
    public Void visitUpdateStatement(UpdateStmt statement, ConnectContext context) {
        boolean hasUpdatePrivOnTargetTable = false;
        ...
        try {
            ...
        } catch (AccessDeniedException e) {
            try {
                Authorizer.checkTableAction(context.getCurrentUserIdentity(), context.getCurrentRoleIds(),
                        statement.getTableName(), PrivilegeType.UPDATE);
                hasUpdatePrivOnTargetTable = true;
            } catch (AccessDeniedException exception) {
                ...
            }
        }
        ...
        Map<TableName, Set<String>> tableColumns = AnalyzerUtils.collectAllSelectTableColumns(statement);
        try {
            checkCanSelectFromColumns(context, tableColumns, allTouchedTables);
        } catch (ErrorReportException e) {
            if (hasUpdatePrivOnTargetTable || !tableColumns.containsKey(statement.getTableName())) {
                allTouchedTables.remove(statement.getTableName());
            }
            checkSelectTableAction(context, allTouchedTables);
        }
```

because with the following sql,
```sql
UPDATE CustPurchase SET id = CI.id 
FROM
   CustID CI 
WHERE
   CI.name = CustPurchase.name
```
If user has **only** `UPDATE` privilege on table `CustPurchase` and `SELECT` privilege on `CustID` , we should let it pass (just as before)?

For `!tableColumns.containsKey(statement.getTableName())` , i think we should also consider sql like `update t set col=xxx` without where predicate, could you confirm that for this case, `col` **won't** go into the result of `AnalyzerUtils.collectAllSelectTableColumns(statement)`? So that we don't need to check the `SELECT` privilege on `t` for this case if we only have `UPDATE` privilege on `t.col`.

And for `update t set t.a=1 where t.b=2`, could you also confirm that `b` **will** go into the result of `AnalyzerUtils.collectAllSelectTableColumns(statement)`?

Reply:
I believe this approach would indeed be better, and I will make the changes accordingly.

For sql like `update t set col=xxx` without where predicate, the result of `AnalyzerUtils.collectAllSelectTableColumns(statement)` would be empty, I will add the test case later.

And for `update t set t.a=1 where t.b=2`, column `b` will go into the result of `AnalyzerUtils.collectAllSelectTableColumns(statement)`, it could be confirmed in `AnalyzeUtilTesttestUpdateStatementCollectColumns.`

Reply:
Exellent, but i forgot another case `update t set t.a=1 where t.a=2`, i.e. the assignment and predicate both use `t.a` as the column, if we **only** have `UPDATE` privilege on `t.a`, we should let it pass or not? If yes, so `AnalyzerUtils.collectAllSelectTableColumns(statement)` in this case should also return empty result? I'm not sure whether your `UpdateStmtSelectTableColumnCollector` has consider this situation. @DaiXinxuan

Reply:
I'm afraid not, it seems that we should remove the columns both used by UPDATE and SELECT in `tableColumns` when execute `checkCanSelectFromColumns`, I'll do it later.

Reply:
> Exellent, but i forgot another case `update t set t.a=1 where t.a=2`, i.e. the assignment and predicate both use `t.a` as the column, if we **only** have `UPDATE` privilege on `t.a`, we should let it pass or not? If yes, so `AnalyzerUtils.collectAllSelectTableColumns(statement)` in this case should also return an empty result? I'm not sure whether your `UpdateStmtSelectTableColumnCollector` has consider this situation. @DaiXinxuan

And this problem could happen in InsertStmt also